upd: update pybind11 and Kokkos version, fix workflows

IvanGrigorik opened the pull request:
This PR contains a basic auto-update for pybind from CMakeLists and updated workflows.
With these changes, I successfully installed PyKokkos on both my local machine and server.

JBludau commented:
Same comment as in #68 : we should work on a strategy to build a newer version of Kokkos. But this should work for the current version

JBludau commented on .github/workflows/linux-ci.yml:
```diff
@@ -30,12 +30,12 @@ jobs:
-      if: ${{ matrix.python-version != 'pypy-3.6' && matrix.python-version != 'pypy-3.7' }}
       run:
         git clone -b ${{ matrix.kokkos-branch }} https://github.com/kokkos/kokkos.git /tmp/kokkos-source &&
+        git reset --hard 15dc143e5f39949eece972a798e175c4b463d4b8 &&
```
No way

JBludau replied:
If you want to do something, then pin the kokkos version and don't use master and develop.

IvanGrigorik commented:
@JBludau It seems like some CI is stuck for 23 hours. Please, cancel it (I don't have enough rights)

IvanGrigorik commented:
I updated Python versions and dependencies from Kokkos 4.4 to Kokkos 4.7
Do not enable CI runs; they still will fail. I'm trying to figure out the issue.

The issue log (`errata.out`) is added to the latest commit just to keep track of errors. Will be removed in the future

kennykos commented on CMakeLists.txt:
```diff
@@ -17,6 +18,23 @@ ENDIF()
+ FETCHCONTENT_DECLARE(
+   Kokkos
+   GIT_REPOSITORY https://github.com/kokkos/kokkos.git
+   GIT_TAG develop
```
Would it make sense to have end users build against master? We should still test against develop, but I think the default fetch should be master.

JBludau commented on .github/workflows/linux-ci.yml:
```diff
@@ -30,12 +30,12 @@ jobs:
-      if: ${{ matrix.python-version != 'pypy-3.6' && matrix.python-version != 'pypy-3.7' }}
       run:
         git clone -b ${{ matrix.kokkos-branch }} https://github.com/kokkos/kokkos.git /tmp/kokkos-source &&
+        git reset --hard 15dc143e5f39949eece972a798e175c4b463d4b8 &&
```
this is still not a good idea

IvanGrigorik replied:
Removed

JBludau commented on include/pools.hpp:
```diff
@@ -50,8 +50,8 @@
-void generate_pool(py::module &_mod, const std::string &_name,
-                   const std::string &_msg) {
+void generate_pool(py::module& _mod, const std::string& _name,
+                   const std::string& _msg) {
```
why?

IvanGrigorik commented on errata.out:
```diff
@@ -0,0 +1,3 @@
+nvlink fatal   : Could not open input file '/home/sifi/.conda/envs/pyk/lib/python3.11/site-packages/pykokkos_base-0.0.7-py3.11-linux-x86_64.egg/lib/libkokkoscontainers.so'
```
I will remove this file as soon as I fix this issue. For now, it just demonstrates what's wrong with this PyKokkos version
For some reason, it can not find `libkokkoscontainers.so`. I am still working on it.

IvanGrigorik commented on .github/workflows/linux-ci.yml:
```diff
@@ -226,7 +226,7 @@ jobs:
     - name: Generate CMakeLists.txt
       run: |
-        echo -e "cmake_minimum_required(VERSION 3.16.0 FATAL_ERROR)\nproject(pykokkos-base-as-subproject LANGUAGES CXX)\nadd_subdirectory(kokkos)\nadd_subdirectory(pybind11)\nadd_subdirectory(pykokkos-base)" > ${{github.workspace}}/CMakeLists.txt
+        echo -e "cmake_minimum_required(VERSION 3.31.0 FATAL_ERROR)\nproject(pykokkos-base-as-subproject LANGUAGES CXX)\nadd_subdirectory(kokkos)\nadd_subdirectory(pybind11)\nadd_subdirectory(pykokkos-base)" > ${{github.workspace}}/CMakeLists.txt
```
Fixed to 3.22

IvanGrigorik commented on .github/workflows/linux-ci.yml:
```diff
@@ -98,16 +98,16 @@ jobs:
           -DKokkos_ENABLE_SERIAL=ON
           -DKokkos_ENABLE_OPENMP=ON
-          -DCMAKE_CXX_STANDARD=17
+          -DCMAKE_CXX_STANDARD=20
```
Rolled back

JBludau commented on CMakeLists.txt:
```diff
@@ -1,6 +1,7 @@
 CMAKE_MINIMUM_REQUIRED(VERSION 3.16 FATAL_ERROR)
 
 SET(CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake/Modules ${CMAKE_MODULE_PATH})
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
```
isn't there a better way? https://cmake.org/cmake/help/latest/prop_tgt/POSITION_INDEPENDENT_CODE.html

JBludau commented on CMakeLists.txt:
```diff
@@ -17,6 +18,23 @@ ENDIF()
 
 INCLUDE(KokkosPythonUtilities)  # miscellaneous macros and functions
 
+# fetch latest stable pykokkos and pybind versions
```
did you mean kokkos and pybind 11?

JBludau commented on include/views.hpp:
```diff
@@ -330,7 +330,8 @@ void generate_view(py::module &_mod, const std::string &_name,
   });
 
-  using mirror_type = typename ViewT::HostMirror;
+  // modern Kokkos uses 'host_mirror_type' (older 'HostMirror' is deprecated)
```
I think we do not need the comment

IvanGrigorik replied:
Fixed

IvanGrigorik commented:
FYI: I finished the Kokkos version update, and everything works on my local laptop, our local server, and inside NVIDIA Docker container. 
I will try to fix workflows now.

IvanGrigorik commented:
`python ./ex-numpy.py` test case works after this change. However, pybind11 throws some errors during testing step:
```
_deps/pybind11-src/tests/conftest.py:25: in <module>
    import pybind11_tests
E   ModuleNotFoundError: No module named 'pybind11_tests'
=============================== short test summary info ===============================
ERROR _deps/pybind11-src/tests - ModuleNotFoundError: No module named 'pybind11_tests'
!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!
================================== 1 error in 0.43s ===================================
```

I don't know if recently removed file could be removed, but I checked and it just duplicated [KokkosExp_InterOp.hpp](https://github.com/kokkos/kokkos/blob/develop/core/src/KokkosExp_InterOp.hpp) file. Due to Kokkos update, this file deprecated and there was problems with pykokkos build. After removing, I checked pykokkos on `ex-numpy.py` and everything works like it described in [readme](https://github.com/kokkos/pykokkos-base?tab=readme-ov-file#expected-output)

IvanGrigorik commented:
I updated PyKokkos and dependency, so now it work if you install it both with cmake of `python setup.py install`
Python CI installation fails due to a different installation process. Installing pykokkos-base from this branch will ensure all pykokkos checks pass.
CIs are failing because of deprecated pip build and subprojects.
@JBludau, can you check changes?

JBludau commented on cmake/Templates/KokkosExp_InterOp.hpp:
```diff
@@ -1,101 +1,23 @@
+#include <Kokkos_Layout.hpp>
+#include <Kokkos_MemoryTraits.hpp>
+#include <Kokkos_View.hpp>
+#include <impl/Kokkos_Utilities.hpp>
```
I am not a big fan of copying the file. But I guess that is a thing for another time

JBludau reviewed (commented):
looks good but we should either get the other install methods to work or we should ditch them.

IvanGrigorik commented on CMakeLists.txt:
```diff
@@ -17,14 +18,30 @@ ENDIF()
+  URL https://github.com/pybind/pybind11/archive/refs/heads/v3.0.zip
+  URL_HASH SHA256=1816ea10bf7de362fb7864b131cc90bed6fc393feaf7e27b0ed5a03987bedd10
+)
+FETCHCONTENT_MAKEAVAILABLE(pybind11)
```
We can check for it as well. Fixed for [pybind11](https://pybind11.readthedocs.io/en/latest/cmake/index.html) and [Kokkos](https://kokkos.org/kokkos-core-wiki/get-started/integrating-kokkos-into-your-cmake-project.html#supporting-both-external-and-embedded-kokkos).

IvanGrigorik commented:
> looks good but we should either get the other install methods to work or we should ditch them.

Actually, most important installation methods works (standard `python setup.py` and `cmake`). There is a problem with deprecated example. 
Milos suggested to remove it and some installation methods that does not work.

IvanGrigorik commented:
## Ready to merge.
All non-formatting/flakyness related workflows are finally up

Formatting and linting workflows would be pushed in the separate PR, since they require some format changes due to clang-format update.

kennykos commented on cmake/Modules/KokkosPythonKokkos.cmake:
```diff
@@ -156,7 +156,22 @@ IF(_INTERNAL_KOKKOS)
+        MESSAGE(STATUS "Fetching Kokkos via FetchContent")
+        FETCHCONTENT_DECLARE(
+          Kokkos
+          URL https://github.com/kokkos/kokkos/archive/refs/heads/release-candidate-4.7.01.zip
```
Will we have to change this with every new release? If so, is there a more automated way to do this?

IvanGrigorik replied:
Nope. We just want to fetch content of some stable version, is external Kokkos was not found. We can fetch from dev/master branches, but currently Kokkos 5.0 is not supported

kennykos reviewed (approved):
This looks good to me! I made some comments about linking directly to kokkos 4.7.1 and updating the README. W.r.t. the README, I support adding badges for our workflows for accountability 😄 and also updating the install instructions.

Of course, we should merge this sooner rather than latter, so maybe these are things addressed in a different PR.

JBludau commented on .github/workflows/linux-ci.yml:
```diff
@@ -191,13 +189,13 @@ jobs:
+        python -m kokkos.test
 
   subproject:
     runs-on: ubuntu-latest
```
We should probably pin the versions ... otherwise we will have silently changing compilers and dependencies

IvanGrigorik replied:
You mean we need to explicitly specify python/kokkos version?
If yes, I will do it in another PR

IvanGrigorik commented on .github/workflows/linux-ci.yml:
```diff
@@ -70,9 +69,21 @@ jobs:
     - name: Test
       working-directory: ${{github.workspace}}/build
       run:
+        export PYTHONPATH=${PWD}/_deps/pybind11-src/build/tests:${PYTHONPATH} &&
```
If we will not add it - pytest will try to execute pynind11 tests and will fail

JBludau replied:
then we should exclude these tests...but right now we are doing cmake's job and rely on cmake internal paths to do this

IvanGrigorik commented on .github/workflows/linux-ci.yml:
```diff
@@ -215,18 +213,21 @@ jobs:
       run: |
         cp -r ${{github.workspace}}/pykokkos-base/external/kokkos ${{github.workspace}}/kokkos
         cp -r ${{github.workspace}}/pykokkos-base/external/pybind11 ${{github.workspace}}/pybind11
+        # Remove .git files/dirs from copied submodules to avoid git errors
```
To make sure that pykokkos can be integrated as cmake project (?)

IvanGrigorik replied:
This comment is outdated. 
I removed git submodules and now we don't need to care about this at all.

IvanGrigorik commented:
> On a second thought: I don't understand some things, and I think we should just remove the submodule ... This is just a worse way to fetch something compared to fetch_content in my opinion

Agree.  I removed all submodules and updated workflows (add `--ignore` flag for deps, updated `subproject` workflow)

@JBludau can you check if we can merge it now?

JBludau commented:
> @gliga told me to merge it and I frankly agree with that We can address all of the comments/suggestions in other PR if we still have any

I fail to see why we rush this but fine.
